### ISSUE TITLE
Notes of expired transaction should be removed

### ISSUE THREAD

**Issue by hairtail** [iron-fish/ironfish#2345]
### What happened?

Incorrect pending balance. When we create a new transaction, we use account.syncTransaction to update notes in db. However, for transactions that are not mined before expiration sequence, something unexpected happens.

```
Your balance is made of notes on the chain that are safe to spend
Balance: $IRON 0.79998400

0 notes worth 0.00000000 are on the chain within 12 blocks of the head
Unconfirmed: 0.79998400

1 notes worth 1.00000000 are waiting for miners to add them to the chain
Pending: 1.79998400

 Status     Creator Hash                                   Miner Fee Fee ($IRON)         Notes Spends Expiration Sequence
 ────────── ─────── ────────────────────────────────────── ───────── ─────────────────── ───── ────── ───────────────────
 confirmed  ✔       9af35cbe06f45d58dea9720eb4366ddbc706d… x         0.00000800          2     1      227327
 confirmed  x       ea0d7631889ce0e934cd9bd2bba399931fe0a… x         0.00005000          2     1      227089
 expired    x       569c8aa82142768410c36ef585c12ba8641c0… x         0.00000500          2     1      227082
 unconfirmed ✔       e6d1d70085f0b1afce89cbf19e0e491a08fe… x         0.00000800          2     1      227356

 Amount ($IRON) Memo       From Transaction                                                 Spent
 ────────────── ────────── ──────────────────────────────────────────────────────────────── ─────
 1.00000000     test 9246cccd66674f2b60740e36a499debdc825f3d3e0ffe040054791db0dd3941a x (from expired transaction)
 0.79998400          b59b1daa28943bfef19b75a3203bd6a89183c37778984d9dafb2eea1c051dfb4 x
 0.89999200          204640b16924cf0c1349a74c1f4caf81d9f62939e4f07908c2478db32d90242c ✔
 1.00000000     test2 4fa27b50544ed0caf63894d40869ec678bf64cba2ebcda3ebee7942fb93420b0 ✔
```

### Version

0.1.49 @ src

### Graffiti

miningKing

**Comment by hairtail** [comment]
`pendingTransactionHashes` of walletdb is empty, but the notes of expired transaction are alive.

**Comment by hughy** [comment]
@hairtail was this example from a transaction created with changes from #2344 applied?

I'm able to reproduce if I make that change, but see notes from expired transactions removed otherwise.

**Comment by hairtail** [comment]
> 4fa27b50544ed0caf63894d40869ec678bf64cba2ebcda3ebee7942fb93420b0

All results shown above are from master branch. #2344 has not been applied yet.

**Comment by hughy** [comment]
Are you able to reproduce this? Can you list the steps to reproduce?

**Comment by hairtail** [comment]
I have tried, but all transactions are mined on chain now. I will try later.

**Comment by hairtail** [comment]
> Are you able to reproduce this? Can you list the steps to reproduce?

A new transaction with fee=500 to send some token to this account, once this transaction expired what shown above happened. There may be one or more node restart.

**Comment by hairtail** [comment]
I use supervisor to manager node process, so there should be a signal for shutting down.

**Comment by hairtail** [comment]
Notes are already spent but marked as unspent. This is based on the latest staging branch.
```
 Amount ($IRON) Memo                             From Transaction                                                 Spent
 ────────────── ──────────────────────────────── ──────────────────────────────────────────────────────────────── ─────
 0.49997598                                      81b811c0673a5425f1eda080d46be7f5e28b2765b55ca9f683b8874da1d56339 ✔
 0.59997599                                      c792af5dfea7ea2b6552b54578b67fcbfbbaad693a403faa9eb56f78e7791ebb x (already spent)
 1.00000000     test                       569c8aa82142768410c36ef585c12ba8641c0bfd5143f3891b6032838d466114 x (from expired transaction)
 0.79998400                                      e6d1d70085f0b1afce89cbf19e0e491a08fe6893aa5fd99c5f2708d9c8f035bb ✔
 0.39997597                                      37e6bcc7afbb32751b68f7472c89d5529046154bb3c3f8b51b3fe39c5c2a2d13 x
 0.69997600                                      ae8ba23b6ad38bbce9b671b06d0b4b4d939260128e2a53763d84a5a60f8bded8 ✔
 0.89999200                                      9af35cbe06f45d58dea9720eb4366ddbc706d9c59efe88c366f1d3b66d18fd55 ✔
 1.00000000     test2                      ea0d7631889ce0e934cd9bd2bba399931fe0aa16fb23dcedd605e1c68435834b ✔
```

```
 Status     Creator Hash                                   Miner Fee Fee ($IRON)         Notes Spends Expiration Sequence
 ────────── ─────── ────────────────────────────────────── ───────── ─────────────────── ───── ────── ───────────────────
 confirmed  ✔       9af35cbe06f45d58dea9720eb4366ddbc706d… x         0.00000800          2     1      227327
 confirmed  x       ea0d7631889ce0e934cd9bd2bba399931fe0a… x         0.00005000          2     1      227089
 expired    ✔       8f76223366ed3b020d48a1006e04435078bea… x         0.00000001          2     1      230374
 expired    x       569c8aa82142768410c36ef585c12ba8641c0… x         0.00000500          2     1      227082
 confirmed  ✔       c792af5dfea7ea2b6552b54578b67fcbfbbaa… x         0.00000001          2     1      230322
 confirmed  ✔       e6d1d70085f0b1afce89cbf19e0e491a08fe6… x         0.00000800          2     1      227356
 expired    ✔       37e6bcc7afbb32751b68f7472c89d55290461… x         0.00000001          2     1      230394
 confirmed  ✔       ae8ba23b6ad38bbce9b671b06d0b4b4d93926… x         0.00000800          2     1      227512
 confirmed  ✔       81b811c0673a5425f1eda080d46be7f5e28b2… x         0.00000001          2     1      230379
```

**Comment by hairtail** [comment]
Status of notes already spent is first marked as spent when I create a transaction to spend it, then it changes and rolls back to unspent.

**Comment by hughy** [comment]
> Status of notes already spent is first marked as spent when I create a transaction to spend it, then it changes and rolls back to unspent.

That's very helpful, thanks for all of the details in the comments here!

**Comment by hughy** [comment]
Resolved in #2399
